Updating docs (#34)
cobaltspeech/sdk-cubic: docs-src/content/using-cubic-sdk/client-configs.md

------

Another use case would be getting both the formatted and raw transcript.  This can be done using this config.

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": "true"
}
```

The formatted transcript would be found at `results.alternatives[*].transcript`, and the raw transcript would be found at `results.alternatives[*].raw_transcript`.

------

If you need to know the timestamp for each word, to align subtitles with a video, for example, then you can use this config to enable those word-level timestamps.  Please note the inclusion of `enable_raw_transcript`; the word-level information corresponds to the raw transcript, since the formatter may combine multiple words into one symbol in the formatted transcript (e.g. "twenty one" to "21")

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": true,
    "enable_word_time_offsets": true
}
```

Timestamps will be found at `results.alternatives[*].words[*].start_time` and `results.alternatives[*].words[*].duration`.

------

> Word-level confidences, i.e. for displaying a lighter color for less confident words, can be include much like the word-level timestamps.  Please note the inclusion of `enable_raw_transcript`.

Reviewer: ```suggestion
Word-level confidences, i.e. for displaying a lighter color for less confident words, can be included much like the word-level timestamps.  Please note the inclusion of `enable_raw_transcript`.
```

---
title: "Recognition Configurations"
weight: 25
---

An in-depth explanation of the various fields of the complete SDK can be found [here](../../protobuf/autogen-doc-cubic-proto/).  The sub-section [RecognitionConfig](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfig) is particularly important here.

This page here discusses the more common combinations sent to the server.

## Fields

> Here's a quick refresher of the fields.

Reviewer 2: "refresher" implies the reader should already know about the fields. Is this the case? (I don't know if there's another page describing the configs that the reader will see before this.) If not, change it to "quick overview".

| Field | Required | Default | Description |
| ----- | -------- | ------- | ----------- |
| model_id | Yes | | Unique ID of the model to use. |
| audio_encoding | Yes | | Encoding format of the audio, such as RAW_LINEAR_16, WAV, MP3, etc. |
| idle_timeout | No | `0s` (Unlimited) | Maximum time allowed between each gRPC message. The server may place further restrictions depending on its configuration. |

> | enable_word_time_offsets | No | `false` | Toggles the calculation of word-level timestamps.  Cubic models must support word level timestamps for this field to be populated. |

Reviewer: ```suggestion
| enable_word_time_offsets | No | `false` | Toggles the calculation of word-level timestamps.  The specified model must also support word level timestamps for this field to be populated. |
```

> | enable_word_confidence | No | `false` | Toggles the calculation of word-level confidence scores. Cubic models must support word level confidence for this field to be populated.|

Reviewer: ```suggestion
| enable_word_confidence | No | `false` | Toggles the calculation of word-level confidence scores. The specified model must also support word level confidence for this field to be populated.|
```

| enable_raw_transcript | No | `false` | If `true`, the raw transcript will be included in the results. |

> | enable_confusion_network | No | `false` | Toggles the creation of a confusion network, consisting of multiple alternatives transcriptions.  Cubic models must support confusion networks for this field to be populated. |

Reviewer: ```suggestion
| enable_confusion_network | No | `false` | Toggles the inclusion of a confusion network, consisting of multiple alternative transcriptions.  The specified model must also support confusion networks for this field to be populated. |
```

| audio_channels | No | `[0]` (mono) | Specifies which channels of a multi-channel audio file to be transcribed, each as their own individual audio stream. |

## Use cases

The most basic use case is getting a formatted transcript.  This would simply need a config such as:

```json
{
    "model_id": "1",
    "audio_encoding": "raw"
}
```

and your resulting transcript would be found at `results.alternatives[*].transcript`.

> Note: Prior to `cubicsvr` v2.9.0 and SDK-Cubic v1.3.0, the field `results.alternatives[*].transcript` was populated with either the raw or the formatted transcription depending on the `enable_raw_transcript` config.  After these changes, raw transcripts have been pushed to a new field `results.alternatives[*].raw_transcript`, and only populated when `enable_raw_transcript` is set to true.

------

> Sometimes, only the Raw transcript is desired.  In this case, there are two options.  

Reviewer: ```suggestion
Sometimes, only the raw (unformatted) transcript is desired.  In this case, there are two options.  
```

1. Disable server-side formatting and use the above config.  Retrieve the raw transcript from the `results.alternatives[*].transcript` field.
2. Specify the `enable_raw_transcript = true` flag, and access the field `results.alternatives[*].raw_transcript`.

> > Note: Prior to `cubicsvr` v2.9.0 and SDK-Cubic v1.3.0, the field `results.alternatives[*].transcript` was populated with either the raw or the formatted transcription depending on the `enable_raw_transcript` config.  After these changes, raw transcripts have been pushed to a new field `results.alternatives[*].raw_transcript`, and only populated when `enable_raw_transcript` is set to true.

Reviewer 2: Do we need this notice duplicated so close to the first notice?

Reviewer: I don't think so.  I think we should have it only here and remove the previous one.

------

Another use case would be getting both the formatted and raw transcript.  This can be done using this config.

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": "true"
}
```

The formatted transcript would be found at `results.alternatives[*].transcript`, and the raw transcript would be found at `results.alternatives[*].raw_transcript`.

------

> If you needed to dive deeper in word level timestamps, say for subtitles, then you can use this config to enable those word-level timestamps.  Please note the inclusion of `enable_raw_transcript`.

Reviewer 2: Be consistent with "word level" or "word-level". I think "word-level" is better.

> If you needed to dive deeper in word level timestamps, say for subtitles, then you can use this config to enable those word-level timestamps.  Please note the inclusion of `enable_raw_transcript`.

Reviewer: ```suggestion
If you need to know the timestamp for each word, to align subtitles with a video, for example, then you can use this config to enable those word-level timestamps.  Please note the inclusion of `enable_raw_transcript`; the word-level information corresponds to the raw transcript, since the formatter may combine multiple words into one symbol in the formatted transcript (e.g. "twenty one" to "21").
```
For Exxon, we implemented providing word-level info for the formatted output.  Is that only implemented for them or should that be an option in cubicsvr directly?  I don't know if there's a pressing need for it, I'm just wondering.

Author: I don't see why we wouldn't support it for everyone, but that's just me.  I'm not sure how we want to handle the format of things though, now that we support raw and formatted side by side.

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": true,
    "enable_word_time_offsets": true
}
```

Timestamps will be found at `results.alternatives[*].words[*].start_time` and `results.alternatives[*].words[*].duration`.

------

> If you needed to dive deeper in word level confidences, i.e. for displaying a lighter color for less confident words, then you can use this config to enable those word-level confidence values.  Please note the inclusion of `enable_raw_transcript`.

Reviewer 2: This line sounds too copy-and-pasted from the description above for enabling word-level timestamps.

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": true,
    "enable_word_time_offsets": true
}
```

Word-level confidence scores will be found at `results.alternatives[*].words[*].confidence`.
 
------

> For power users, such as those combining results from multiple ASR engines, the most comprehensive and detailed Cubic results are found in the confusion network.  Please refer to the [in depth confusion network documetation](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfusionnetwork) for more details.

Reviewer 2: What is a "power user"?

Reviewer: I agree that "power user" doesn't quite strike the right tone.   Perhaps something like 

```suggestion
For applications that need more than the one-best transcript, the most comprehensive and detailed Cubic results are found in the confusion network.  Please refer to the [in depth confusion network documentation](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfusionnetwork) to see what is included.
```

> For power users, such as those combining results from multiple ASR engines, the most comprehensive and detailed Cubic results are found in the confusion network.  Please refer to the [in depth confusion network documetation](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfusionnetwork) for more details.

Reviewer 3: Also, I think it would make more sense to link to the confusion network documentation on SDK website homepage (there isn't much of a description in the protobuf documentation). I *think* the hyperlink is something like `../../#confusion-network`, but I'm not entirely sure how Hugo translates these from markdown.

---
title: "Recognition Configurations"
weight: 25
---

> An indepth explination of the various fields of the complete SDK can be hound [here](../../protobuf/autogen-doc-cubic-proto/).  The sub-section [RecognitionConfig](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfig) is particularly important here.

Reviewer 2: ```suggestion
An in-depth explanation of the various fields of the complete SDK can be found [here](../../protobuf/autogen-doc-cubic-proto/).  The sub-section [RecognitionConfig](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfig) is particularly important here.
```

| enable_raw_transcript | No | `false` | Enables the raw transcript to be included or not. |
| enable_confusion_network | No | `false` | Toggles the creation of a confusion network, consisting of multiple alternatives transcriptions.  Cubic models must support confusion networks for this field to be populated. |
| audio_channels | No | `[0]` (mono) | Allows multi-channel audio files to be transcribed as individual audio streams |

## Use cases

The most basic use case is getting a formatted transcript.  This would simply need a config such as:

```json
{
    "model_id": "1",
    "audio_encoding": "raw"
}
```

and your resulting transcript would be found at `results.alternatives[*].transcript`.

> Note: Prior to CubicSvr v2.9.0 and SDK-Cubic v1.3.0, the field `results.alternatives[*].transcript` was populated with either the raw or the formatted transcription depending on the `enable_raw_transcript` config.  After these changes, raw transcripts have been pushed to a new field `results.alternatives[*].raw_transcript`, and only populated when `enable_raw_transcript` is set to true.

------

Sometimes, only the Raw transcript is desired.  In this case, there are two options.  

> 1) Disable Server-side formatting, and using the above config and the `results.alternatives[*].transcript` field.

Reviewer 3: ```suggestion
1. Disable server-side formatting and use the above config. Retrieve the raw transcript from the `results.alternatives[*].transcript` field.
```

> 2) Include the `enable_raw_transcript = true` flag, and access the field `results.alternatives[*].raw_transcript`.

Reviewer 3: ```suggestion
2. Specify the `enable_raw_transcript = true` flag, and access the field `results.alternatives[*].raw_transcript`.
```

> formatting If you use `enable_raw_trnascript`

Reviewer 3: ```suggestion
```

> > Note: Prior to CubicSvr v2.9.0 and SDK-Cubic v1.3.0, the field `results.alternatives[*].transcript` was populated with either the raw or the formatted transcription depending on the `enable_raw_transcript` config.  After these changes, raw transcripts have been pushed to a new field `results.alternatives[*].raw_transcript`, and only populated when `enable_raw_transcript` is set to true.

Reviewer 3: ```suggestion
```
This repeats the same note as above (line 38).

Author: That was intentional.

------

> Another use case would be getting both the formatted and raw transcript.  If this was needed, you could use the config

Reviewer 3: ```suggestion
Another use case would be getting both the formatted and raw transcript.  This can be done using the config
```

    "enable_raw_transcript": "true"
}
```

The formatted transcript would be found at `results.alternatives[*].transcript`, and the raw transcript would be found at `results.alternatives[*].raw_transcript`.

------

If you needed to dive deeper in word level timestamps, say for subtitles, then you can use this config to enable those word-level timestamps.  Please note the inclusion of `enable_raw_transcript`.

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": true,
    "enable_word_time_offsets": true
}
```

Timestamps will be found at `results.alternatives[*].words[*].start_time` and `results.alternatives[*].words[*].duration`.

------

If you needed to dive deeper in word level confidences, i.e. for displaying a lighter color for less confident words, then you can use this config to enable those word-level confidence values.  Please note the inclusion of `enable_raw_transcript`.

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": true,
    "enable_word_time_offsets": true
}
```

Word-level confidence scores will be found at `results.alternatives[*].words[*].confidence`.
 
------

> For power users, such as those combining results from multiple different ASR engines, the deepest level of details are found in the Confusion network.  Please refer to the [in depth Confusion Network documetation](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfusionnetwork) for more details about the contents.

Reviewer 3: ```suggestion
For power users, such as those combining results from multiple ASR engines, the most comprehensive and detailed Cubic results are found in the confusion network.  Please refer to the [in depth confusion network documentation](../../protobuf/autogen-doc-cubic-proto/#message-recognitionconfusionnetwork) for more details.
```

> The config used would look something like this:

Reviewer 3: ```suggestion
To enable the confusion network, the config will look similar to this:
```

```json
{
    "model_id": "1",
    "audio_encoding": "raw",
    "enable_raw_transcript": true,
    "enable_confusion_network": true
}
```

> The confusino network can be accessed at `result.cnet`.

Reviewer 3: ```suggestion
The confusion network can be accessed at `result.cnet`.
```
